app/src/main/java/com/example/dungeontest/graph/OpenAiRequestBuilder.kt

```diff
@@ -0,0 +1,133 @@
+        Third, please make a JSON array describing that graph (root level key for the array is "final_result"), where each entry in the array is a json object containing a) a unique numeric node id (json key "id"), b) the room label from the drawing (json key "label"), and c) the id's of the rooms which it's connected to by hallways (json key "neighbors").
+    """.trimIndent()
+
+    private val client = OkHttpClient
+        .Builder()
+        .connectTimeout(40, TimeUnit.SECONDS)
+        .readTimeout(400, TimeUnit.SECONDS)
+        .build()
+    /* Probably can adjust this? */
+    private val maxTokenCount = 3000
+
+    fun sendOpenAIRequest(apiKey: String, apiVersion: String, imagePath: String, callback: (String?) -> Unit) {
+        val imageFile = File(imagePath)
+        if (!imageFile.exists()) {
+            println("Image file not found: $imagePath")
+            callback(null)
+            return
+        }
+
+        val imageBytes = imageFile.readBytes()
+        Log.v("OpenAiRequestBuilder", imageBytes.size.toString())
+        if (imageBytes.size > 20 * 1024 * 1024) {
+            println("Image file is too large: $imagePath")
+            callback(null)
+            return
+        }
+
+        val supportedFormats = listOf("png", "jpeg", "gif", "webp")
+        val fileExtension = imageFile.extension.lowercase()
+        if (fileExtension !in supportedFormats) {
+            println("Unsupported image format: $fileExtension")
+            callback(null)
+            return
+        }
+
+        val imageBase64 = Base64.getEncoder().encodeToString(imageBytes)
+        val dataUri = "data:image/$fileExtension;base64,$imageBase64"
+
+        val jsonBody = JSONObject().apply {
+            put("model", apiVersion)
+            put("response_format", JSONObject().apply {
+                put("type", "json_object")
+            })
+            put("messages", JSONArray().apply {
+                put(JSONObject().apply {
+                    put("role", "user")
+                    put("content", JSONArray().apply {
+                        put(JSONObject().apply {
+                            put("type", "text")
+                            put("text", prompt)
+                        })
+                        put(JSONObject().apply {
+                            put("type", "image_url")
+                            put("image_url", JSONObject().apply {
+                                put("url", dataUri)
+                            })
```

Comment:
I think the previous 3 lines can be simplified to one:

`put("image_url", JSONObject().put("url", dataUri))`

please let me know if I'm misunderstanding the use of apply {} here

Reply:
View above comment